src/rascal/math/kvec_generator.hh

```diff
@@ -0,0 +1,229 @@
+         * We need to distinguish whether the provided vectors are already
+         * reciprocal cell vectors or need to be converted first
+         */
+        if (not is_reciprocal_cell) {  // real space cell is given
+          // Create reciprocal cell from real space cell
+          Eigen::Matrix3d tcell = basisvectors.transpose();
+          this->basisvecs = 2.0 * M_PI * tcell.inverse();
+        } else {  // cell of reciprocal space is already given
+          this->basisvecs = basisvectors;
+        }
+
+        // Run the precompute function described below
+        this->precompute();
+      }  // end constructor
+
+      /**
+       * Using the provided parameters in the constructor, find
+       * all the k-vectors within the cutoff radius and store them
+       * as well as derived quantities that can be returned to the user.
+       */
+      void precompute();
+
+      // FUNCTIONS FOR USER
+
+      /** Returns number of vectors found within cutoff radius
+       * without double counting pairs related by inversion.
+       * This is the actual number that can be used to iterate over all
+       * vectors and its norms obtained from get_kvectors and get_norms.
+       */
+      size_t get_numvectors() const { return this->numvectors; }
+
+      /** Returns number of vectors found within cutoff radius
+       * counting pairs related by inversion separately.
+       * If the origin is excluded, the number without origin is doubled
+       * and returned. Otherwise, the origin is counted once.
+       */
+      size_t get_numvectors_all() const {
+        if (this->include_origin) {
+          return 2 * this->numvectors - 1;
+        } else {
+          return 2 * this->numvectors;
+        }
+      }
+
+      /** Returns the Eigen matrix of shape this->numvectors x 3 containing
+       * all k-vectors within cutoff radius without double counting,
+       * where row(i) = i-th vector. If the origin k=(0,0,0) is included,
+       * it is always stored in row(0).
+       */
+      Matrix_t get_kvectors() const { return this->kvecs; }
+
+      /** Returns the Eigen vector of length this->numvectors containing
+       * the norm of all k-vectors within the cutoff without double counting.
+       */
+      Vector_t get_norms() const { return this->kvecnorms; }
+
+      // FUNCTIONS FOR DEVELOPERS / TESTING
+
+      /**
+       * Return the expected number of obtained vectors based on
+       * continuous approximation, which is simply the volume of a ball
+       * of radius kcut divided by the reciprocal space unit cell.
+       * In practice, we have to divide by 2 since we are only generating
+       * half of the k-vectors, which is why we use 2pi/3 rather than 4pi/3.
+       * The volume of the unit cell in reciprocal space is sqrt(detM).
+       */
+      double get_numvectors_estimate() const {
+        const double kcut = this->kcutoff;
+        return 2. * M_PI / 3. * kcut * kcut * kcut / sqrt(this->detM);
```

Comment:
I don't know why, but we have defined our own pi constant in math/utils.hh I would just use it here for consistency within the project